netmap/netmap_contract.go

```diff
@@ -457,17 +441,101 @@ func NetmapCandidates() []netmapNode {
 // Netmap contract contains only two recent network map snapshot: current and
 // previous epoch. For diff bigger than 1 or less than 0 method throws panic.
 func Snapshot(diff int) []storageNode {
-	if diff < 0 || SnapshotCount <= diff {
+	ctx := storage.GetReadOnlyContext()
+	count := getSnapshotCount(ctx)
+	if diff < 0 || count <= diff {
 		panic("incorrect diff")
 	}
 
-	ctx := storage.GetReadOnlyContext()
 	id := storage.Get(ctx, snapshotCurrentIDKey).(int)
-	needID := (id - diff + SnapshotCount) % SnapshotCount
+	needID := (id - diff + count) % count
 	key := snapshotKeyPrefix + string([]byte{byte(needID)})
 	return getSnapshot(ctx, key)
 }
 
+func getSnapshotCount(ctx storage.Context) int {
+	return storage.Get(ctx, snapshotCountKey).(int)
+}
+
+// UpdateSnapshotCount updates number of stored snapshots.
+// If new number is less than the old one, old snapshots are removed.
+// Otherwise, history is extended to with empty snapshots, so
+// `Snapshot` method can return invalid results for `diff = new-old` epochs
+// until `diff` epochs have passed.
+func UpdateSnapshotCount(count int) {
+	common.CheckAlphabetWitness(common.AlphabetAddress())
+	if count < 0 {
+		panic("count must be positive")
+	}
+	ctx := storage.GetContext()
+	curr := getSnapshotCount(ctx)
+	if curr == count {
+		panic("count has not changed")
+	}
+	storage.Put(ctx, snapshotCountKey, count)
+
+	id := storage.Get(ctx, snapshotCurrentIDKey).(int)
+	var delStart, delFinish int
+	if curr < count {
+		// Increase history size.
+		//
+		// Old state (N = count, K = curr)
+		// KEY INDEX: 0   | 1     | ... | E | E+1   | ... | K-1   | ... | N-1
+		// EPOCH    : C-E | C-E+1 | ... | C | C-K+1 | ... | C-E-1 |
+		//
+		// New state:
+		// KEY INDEX: 0   | 1     | ... | E | E+1 | ... | K-1 | ... | N-1
+		// EPOCH    : C-E | C-E+1 | ... | C | nil | ... | .   | ... | C-E-1
+		//
+		// So we need to move tail snapshots N-K keys forward,
+		// i.e. from E+1 .. K to N-K+E+1 .. N
+		diff := count - curr
+		lower := diff + id + 1
+		for k := count - 1; k >= lower; k-- {
+			moveSnapshot(ctx, k-diff, k)
+		}
+		delStart, delFinish = id+1, id+1+diff
+	} else {
+		// Decrease history size.
+		//
+		// Old state (N = curr, K = count)
+		// KEY INDEX: 0   | 1     | ... K1 ... | E | E+1   | ... K2-1 ... | N-1
+		// EPOCH    : C-E | C-E+1 | ... .. ... | C | C-N+1 | ... ...  ... | C-E-1
+		var step, start int
+		if id < count {
+			// K2 case, move snapshots from E+1+N-K .. N-1 range to E+1 .. K-1
+			// New state:
+			// KEY INDEX: 0   | 1     | ... | E | E+1   | ... | K-1
+			// EPOCH    : C-E | C-E+1 | ... | C | C-K+1 | ... | C-E-1
+			step = curr - count
+			start = step + id + 1
```

Comment:
I think it should be `start = id + 1` there. You get `+step` in the `for` loop down below. 

In current implementation with this example I get `step = 2`, `start = 5`, therefore `for` loop has 0 iteration.
```
before: [g, h, i, d, e, f]
               ^ - id
after:  [g, h, i, f]  
               ^ - id
```